spcl/dace
# Support for HBM on Xilinx FPGA's (pull-request 2)

Add Support for HBM for Xilinx FPGA's

Changes:

sdfg/utils: 
- Added several utility functions for HBM-arrays/subsets to utils

targets/cpp: 
- Modifed ptr to generate distinct names for HBM-arrays. The function array_interface_variable is only called from ptr.
- Modified several functions in targets/cpp slighly to support HBM-arrays

targets/fpga:
- Modifed several of the functions which generate arguments/definitions and so on for variables to create multiple of those objects in case of HBM-arrays
- make_parameters recognizes and passes HBM and DDR in it's bank_assignments return value
- emit_copy has support for Host2Device, Device2Host, Device2Device and on Device copies from/to HBM-arrays and also between HBM and DDR. Writing directly from HBM to Streams is not yet supported (it can be done using tasklets with single bank memlets)
- copy_memory "disolves" the magic index for HBM-multibank memlets and only calls emit_copy with single bank HBM memlets
- found a possible old bug (maybe this is untested) at line 1423. Altered it to what seems to make sense there.

targets/xilinx:
- Modifed several of the functions which generate arguments/definitions and so on for variables to create multiple of those objects in case of HBM-arrays
- get_generated_codeobjects: HBM interfaces are declared

targets/cpu:
- Modifed calls to cpp.ptr in some locations
- Added NotImplementedException if nested sdfg's on cpu side deal with HBM-arrays. This will likely be supported in the next PR, but at the moment it seems to much like a nice to have.

sdfg/validation:
- Check that no HBM multibank memlets are attached to a tasklet
- Check that the bank/hbmbank location property is valid for the given array (correct type, correct arrayshape, ...)
- Check that all HBM-memlets have constant evaluatable indices

dace/data/*
- Added the value type of the locationproperty to str from pystr_to_symbolic. Needed because of the serialization for ranges

transformation/fpga_transform_state
- A few smaller modifications so that hbmbank is correctly set

tests/fpga/*
- Added several tests for all the mentioned changes above
